File: Minimum-Add-to-Make-Parentheses-Valid/MAtMPV.rs
pub fn min_add_to_make_valid(s: String) -> i32 {
    let mut cache = 0;
    let mut result = 0;
    for c in s.bytes() {
        match c {
            b'(' => cache += 1,
            b')' => {
                if cache == 0 {
                    result += 1
                } else {
                    cache -= 1;
                }
            }
            _ => (),
        }
    }
    result + cache
}

fn main() {
    dbg!(min_add_to_make_valid(String::from("())")));
    dbg!(min_add_to_make_valid(String::from("(((")));
    dbg!(min_add_to_make_valid(String::from("()")));
    dbg!(min_add_to_make_valid(String::from("()))((")));
}

File: Minimum-Add-to-Make-Parentheses-Valid/MAtMPV.hs
min_add_to_make_valid :: [Char] -> Int
min_add_to_make_valid s = loop_with (0, 0) s
  where
    loop_with :: (Int, Int) -> [Char] -> Int
    loop_with (cache,result ) [] = result + cache
    loop_with (cache, result) (x:xs)
      | x == '(' = loop_with ((cache + 1), result) xs
      | x == ')' =
        loop_with
          (if cache == 0
             then (0, (1 + result))
             else (cache - 1, result)) xs
      |otherwise = 0


main = do
  print $ min_add_to_make_valid "())"
